### PR TITLE
Ntp status

### LINKED ISSUE #841: Add status support for NTP

### PR DESCRIPTION
**Pull request by mattiaswal**
## Description
This fix #841 

## Checklist

Tick *relevant* boxes, this PR is-a or has-a:

- [ ] Bugfix
  - [ ] Regression tests
  - [ ] ChangeLog updates (for next release)
- [x] Feature
  - [x] YANG model change => revision updated?
  - [x] Regression tests added?
  - [ ] ChangeLog updates (for next release)
  - [ ] Documentation added?
- [x] Test changes
  - [x] Checked in changed Readme.adoc (make test-spec)
  - [x] Added new test to group Readme.adoc and yaml file
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (please detail in commit messages)
- [ ] Build related changes
- [ ] Documentation content changes
  - [ ] ChangeLog updated (for major changes)
- [ ] Other (please describe):

### CODE REVIEW

**Comment by mattiaswal** [comment]
"show ntp" in CLI has also been added:
<pre>
admin@target:/> show ntp 
ADDRESS         MODE         STATE        STRATUM  POLL-INTERVAL
192.168.1.1     server       candidate    1        7    
192.168.2.1     server       candidate    1        7    
192.168.3.1     server       selected     1        6    
</pre>

File: src/confd/yang/infix-system.yang [hunk 1]
```diff
@@ -115,6 +125,41 @@ module infix-system {
...
+        description "Normal NTP server";
+      }
+      enum peer {
+        description "Symmetric peer";
+      }
+      enum local-clock {
+        description "Reference clock";
+      }
+    }
+}
```
Comment (troglobit) [inline src/confd/yang/infix-system.yang:162]: Is there any reference to the terminology used here? If so, please add a link to RFC and section, for example.  I mostly reacted to "Normal" and "falseticker" ...

File: src/confd/yang/infix-system.yang [hunk 2]
```diff
@@ -256,6 +301,37 @@ module infix-system {
...
+            description "Address to connected NTP server";
+          }
+          leaf source-mode {
+            type source-mode;
+            description "Source mode";
+          }
+          leaf source-state {
+            type source-state;
+            description "Source state";
+          }
```
Comment (troglobit) [inline src/confd/yang/infix-system.yang:321]: As I mentioned in passing IRL, do we need the `source-` prefix for these nodes?  Are we expecting a "target-mode" in the source list?  Maybe just `mode` and `state` are sufficient, with a description similar to "Server mode" and "Server state"?

File: src/confd/yang/infix-system.yang [hunk 3]
```diff
@@ -256,6 +301,37 @@ module infix-system {
...
 
+  augment "/sys:system-state" {
+    container ntp {
+      description "NTP status";
+      container sources {
+        list source {
+          key address;
+          leaf address {
+            type inet:ip-address;
+            description "Address to connected NTP server";
```
Comment (troglobit) [inline src/confd/yang/infix-system.yang:312]: Iirc, it's UDP, so maybe replace/drop "to connected" ... perhaps just "Server address".

File: src/confd/yang/infix-system.yang [hunk 4]
```diff
@@ -256,6 +301,37 @@ module infix-system {
...
+            type source-mode;
+            description "Source mode";
+          }
+          leaf source-state {
+            type source-state;
+            description "Source state";
+          }
+          leaf stratum {
+            type uint8;
+            description "Stratum of connected NTP server";
```
Comment (troglobit) [inline src/confd/yang/infix-system.yang:324]: Same here, drop "connected".

File: test/case/ietf_system/ntp_client/Readme.adoc [hunk 5]
```diff
@@ -0,0 +1,24 @@
+=== Simple NTP client test
+==== Description
+Verify NTP client to multiple NTP servers and verify that one get selected
```
Comment (troglobit) [inline test/case/ietf_system/ntp_client/Readme.adoc:3]: Grammar, maybe "Verify NTP client with multiple servers, ensure one get selected."

Also, what's the basis of the selection, strata?  Then perhaps say "the best one get selected."

File: test/infamy/ntp.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+"""
+NTP client helper
+"""
+def _get_ntp(target):
```
Comment (troglobit) [inline test/infamy/ntp.py:4]: Pylint, two leading blank rows after heading.

File: test/infamy/ntp.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    xpath="/ietf-system:system-state/infix-system:ntp"
+    data=target.get_data(xpath)
```
Comment (troglobit) [inline test/infamy/ntp.py:6]: Pylint: spaces between assignment: `foo = bar`.

File: test/infamy/ntp.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+    if data is None:
+        return None
+
+    return data["system-state"].get("infix-system:ntp", None) or data["system-state"].get("ntp", None)
+
+def _get_ntp_sources(target):
+    ntp=_get_ntp(target)
```
Comment (troglobit) [inline test/infamy/ntp.py:14]: same here

File: test/infamy/ntp.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+    if ntp is None:
+        return []
+
+    return ntp["sources"]["source"]
+
+def any_source_selected(target):
+    sources=_get_ntp_sources(target)
```
Comment (troglobit) [inline test/infamy/ntp.py:22]: and here

**Review by troglobit — CHANGES_REQUESTED** [review]
Some minor comments and a request for clarification.  Otherwise, great work as usual! :muscle:

File: test/case/ietf_system/ntp_client/Readme.adoc (hunk 5 above)
Comment (mattiaswal) [reply test/case/ietf_system/ntp_client/Readme.adoc:3]: All server has same stratum etc, not possible to change in busybox ntpd, so now it just test that "someone" gets selected.

Comment (troglobit) [reply test/case/ietf_system/ntp_client/Readme.adoc:3]: OK, that's fine :smiley:

File: doc/ChangeLog.md [hunk 10]
```diff
@@ -221,7 +225,7 @@ renamed to ease maintenance, more info below.
   with `custom-phys-address` to allow for constructing more free-form
   MAC addresses based on the chassis MAC (a.k.a., base MAC) address.
   For more information, see the YANG model, a few examples are listed in
-  the updated documentation.
```
Comment (troglobit) [inline doc/ChangeLog.md:224]: Double space in markdown to enforce a newline in the output.

File: doc/ChangeLog.md [hunk 11]
```diff
@@ -52,6 +52,10 @@ All notable changes to the project are documented in this file.
    forwarded, when the underlying ports are simultaneously attached to
    a VLAN filtering bridge.
 
+[v24.11.0][] - UNRELEASED
+### Changes
```
Comment (troglobit) [inline doc/ChangeLog.md:56]: Should be

    [v24.12.0][UNRELEASED]
    --------------------------

**Review by troglobit — APPROVED** [review]
Awesome work, approved with only minor comments! :sunglasses:

**Review by wkz — APPROVED** [review]
Spotless! :sparkles:

**Pull request merged**